### ISSUE TITLE
[Proposal] Don't allow to create an entity (class, class's field) if its name consists of one of the keywords

### ISSUE THREAD

**Issue by maseev** [orientechnologies/orientdb#5537]
Hi there. I'm using the latest version of OrientDB at the moment (2.1.8). The problem I've just run into is that you can create for example a class with the Orient studio which name would be ```Cluster```.

Later, if you try to perform a simple query like ``` select from Cluster ``` it'll blow up with an exception which is not very informative. There's the same problem with property creation. You can create a class property named ``` cluster ```, for example. If you try to query an entity like so:

``` select from Entity where cluster = "something" ```

it'll again blow up with an exception.

Maybe, it would be more appropriate to throw an exception at the class or class property creation time (registering an entity and generation the schema) rather than at the time where the actual query is performed?

By the way, [here's](https://github.com/maseev/orient/blob/master/src/test/java/orientdb/OrientDbTest.java#L145) a simple test case which you can run in order to reproduce the problem. Run it with ``` mvn test -Dtest=orientdb.OrientDbTest#registeringABrokenEntityShouldThrowAnException ```

**Comment by luigidellaquila** [comment]
Hi @maseev 

Right now you can name properties with keywords, but you have to quote them with back-ticks in queries, eg.

```
select from Entity where `cluster` = "something"
```

IMHO we should just make this consistent in the CREATE CLASS/PROPERTY statements, eg. this statement should throw an exception
```
CREATE PROPERTY Entity.cluster String
```
but this one should succeed
```
CREATE PROPERTY Entity.`cluster` String
```

This should not be a big effort, I hope I'll be able to put it in 2.2 final

Thanks

Luigi

**Comment by maseev** [comment]
@luigidellaquila Ok, this query

```
select from Entity where `cluster` = "something"
```

is going to work out, but this one
```
CREATE PROPERTY Entity.`cluster` String
```

isn't. You'll get an exception:

```
com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #46: Invalid keyword '`.TYPE' Command: SELECT count(*) FROM `Entity` WHERE `'cluster'`.type() NOT IN ["STRING"] AND `'cluster'` IS NOT NULL ------------------------------------------------------^
```

Moreover, if you have a class property named cluster, you won't be able to add any records to the class. Try to execute this query:

```
insert into Entity (cluster) values ("test")
```

and you'll get an exception:

```
com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Error parsing query: insert into Entity (cluster) values ("test") Encountered " "(" "( "" at line 1, column 20. Was expecting one of: <SELECT> ... <FROM> ... <RETURN> ... <CLUSTER> ...
```

Maybe if I quote a class field it'll work out:

```
insert into Entity ('cluster') values ("test")
```

Nope, no luck again:

```
com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Error parsing query: insert into Entity ('cluster') values ("test") Encountered " "(" "( "" at line 1, column 20. Was expecting one of: <SELECT> ... <FROM> ... <RETURN> ... <CLUSTER> ...
```

Let's just omit the class field altogether and let the OrientDB decide what to do:

```
insert into Entity values ("test")
```

Well, it's not gonna work either:

```
com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Error parsing query: insert into Entity values ("test") Encountered " <VALUES> "values "" at line 1, column 20. Was expecting one of: <EOF> <SELECT> ... <FROM> ... <RETURN> ... <UNSAFE> ... "(" ... ";" ... <CLUSTER> ... "(" ...
```

What I'm trying to say is I think OrientDB absolutely must throw an exception in the first place and not allow users to create classes or class fields whose name can consist of some reserved keywords.

**Comment by luigidellaquila** [comment]
@maseev yes, exactly, that's the part that has to be fixed

Thanks

Luigi

**Comment by maseev** [comment]
Ok. I'd appreciate if you place this ticket to the right milestone so it won't be lost.

**Comment by luigidellaquila** [comment]
OK, I'll tentatively tag it for 2.2

Thanks

Luigi

**Comment by zerovian** [comment]
You can add 'Lock' and 'Index' to this list as well.

**Comment by nagarajasr** [comment]
CREATE PROPERTY Entity.\`cluster\` String already works in 2.2.0-SNAPSHOT:

```sql
orientdb {db=GratefulDeadConcerts}> create class entity

Class created successfully. Total classes in database now: 14.

orientdb {db=GratefulDeadConcerts}> CREATE PROPERTY Entity.`cluster` String

Property created successfully with id=1.

orientdb {db=GratefulDeadConcerts}>
```

so does this:

```
orientdb {db=GratefulDeadConcerts}> insert into entity (abc, `cluster`) values ("pass", "pass")

Inserted record 'entity#30:0{abc:pass,`cluster`:pass} v1' in 0.010000 sec(s).


```

but fails when a single field is being inserted:

```
orientdb {db=GratefulDeadConcerts}> insert into entity (`cluster`) values ("fails")

Error: com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #-1: Cannot find expected keyword '[VALUES]'. Use INSERT INTO [class:]<class>|cluster:<cluster>|index:<index> [(<field>[,]*) VALUES (<expression>[,]*)[,]*]|[SET <field> = <expression>|<sub-command>[,]*]|CONTENT {<JSON>} [RETURN <expression>] [FROM select-query]
Command: INSERT INTO entity (`cluster`) VALUES ("fails")
---------^
        Storage URL="plocal:/home/nagarajasr/src/orientdb/releases/orientdb-community-2.2.0-SNAPSHOT/databases/GratefulDeadConcerts"
        Storage URL="plocal:/home/nagarajasr/src/orientdb/releases/orientdb-community-2.2.0-SNAPSHOT/databases/GratefulDeadConcerts"

```

**Comment by luigidellaquila** [comment]
Hi @nagarajasr 

thank you for reporting, I'll check this specific case

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @nagarajasr 

I just pushed a fix on develop branch

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @nagarajasr @maseev 

All this was fixed in 2.2 GA. I'm closing this issue, please reopen if you see any related problems

Thanks

Luigi
